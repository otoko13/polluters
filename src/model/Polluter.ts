export interface IPolluterArgs {
  rank: number;
  name: string;
  longitude: number;
  latitude: number;
  globalEmissions: string;
  ownership: string;
  hq: string;
  ceo: string | undefined;
  ceoAnnualPay: string | undefined;
  fossilFuelProduction: string | undefined;
  investmentInRenewables: string | undefined;
  globalEmissions1965to2017: string | undefined;
  projectedIncreaseInProduction2018to2030: string | undefined;
  projectedEmissions2018to2030: string | undefined;
  environmentalDisaster: string | undefined;
  environmentalScandal: string | undefined;
  futureProjects: string | undefined;
  revenue: string | undefined;
}

export enum EOwnership{
  State,
  Shareholders,
  StateAndShareholders,
  Unknown,
}

export interface IPolluter extends Polluter {}

const NO_INFORMATTON = 'No information';

class Polluter {
  readonly name: string;
  readonly rank: number;
  readonly longitude: number;
  readonly latitude: number;
  readonly globalEmissions: string;
  readonly ownership: EOwnership;
  readonly hq: string;
  readonly ceo: string;
  readonly annualCeoPay: number | undefined;
  readonly fossilFuelProduction: string;
  readonly investmentInRenewables: string;
  readonly globalEmissions1965to2017: string;
  readonly projectedIncreaseInProduction2018to2030: string;
  readonly projectedEmissions2018to2030: string;
  readonly environmentalDisaster: string | undefined;
  readonly environmentalScandal: string | undefined;
  readonly futureProjects: string;
  readonly revenue: string;

  private static getOwnership(ownership: string) {
    const isState = ownership.toLowerCase().indexOf('state') > -1;
    const isShareholder = ownership.toLowerCase().indexOf('shareholder') > -1;
    if (isState && isShareholder) {
      return EOwnership.StateAndShareholders;
    }
    if (isState) {
      return EOwnership.State;
    }
    if (isShareholder) {
      return EOwnership.Shareholders;
    }
    return EOwnership.Unknown;
  }

  private static getCeoAnnualPay(value: string | undefined) {
    if (!value) {
      return;
    }
    const regEx = /^\$(.+)m(.)*$/gi;
    const payMatch = regEx.exec(value);
    if (!payMatch) {
      return;
    }
    return parseFloat(payMatch[1]);
  }

  constructor(args: IPolluterArgs) {
    this.name = args.name;
    this.rank = args.rank;
    this.longitude = args.longitude;
    this.latitude = args.latitude;
    this.globalEmissions = args.globalEmissions;
    this.hq = args.hq;
    this.ceo = args.ceo || 'Unknown';
    this.fossilFuelProduction = args.fossilFuelProduction || NO_INFORMATTON;
    this.investmentInRenewables = args.investmentInRenewables || NO_INFORMATTON;
    this.globalEmissions1965to2017 = args.globalEmissions1965to2017 || NO_INFORMATTON;
    this.projectedIncreaseInProduction2018to2030 = args.projectedIncreaseInProduction2018to2030 || NO_INFORMATTON;
    this.projectedEmissions2018to2030 = args.projectedEmissions2018to2030 || NO_INFORMATTON;
    this.environmentalDisaster = args.environmentalDisaster;
    this.environmentalScandal = args.environmentalScandal;
    this.futureProjects = args.futureProjects || NO_INFORMATTON;
    this.revenue = args.revenue || NO_INFORMATTON;

    this.annualCeoPay = Polluter.getCeoAnnualPay(args.ceoAnnualPay);
    console.log(this.annualCeoPay);

    this.ownership = Polluter.getOwnership(args.ownership);
  }
}

export default Polluter;
