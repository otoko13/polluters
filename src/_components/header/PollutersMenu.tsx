import React from 'react';

import { Menu, Button, MenuItem } from '@material-ui/core';
import { IPolluter } from '../../model/Polluter';
import './pollutersMenu.scss';

export interface IPollutersMenuProps {
  polluters: IPolluter[];
  onSelected: (polluter: IPolluter) => void;
  onHovered: (polluter: IPolluter) => void;
  onNothingHovered: () => void;
}

const PollutersMenu = (props: IPollutersMenuProps) => {
  const [menuOpen, setMenuOpen] = React.useState(false);
  const menuAnchorRef = React.useRef<HTMLDivElement>(null);

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    setMenuOpen(true);
  };

  const handleClose = () => {
    setMenuOpen(false);
  };

  function handleSelected(polluter: IPolluter) {
    setMenuOpen(false);
    props.onSelected(polluter);
  }

  function getMenuItems() {
    return props.polluters
      .sort((p1, p2) => (p1.rank > p2.rank ? 1 : -1))
      .map(p => (
        <MenuItem
          key={p.rank}
          onClick={() => handleSelected(p)}
          onMouseOver={() => props.onHovered(p)}
          onMouseOut={() => props.onNothingHovered()}
          onFocus={() => props.onHovered(p)}
          onBlur={() => props.onNothingHovered()}
        >
          {p.rank}. {p.name}
        </MenuItem>
      ));
  }

  return (
    <div className="PollutersMenu">
      <Button 
        variant="contained" 
        aria-controls="polluters-menu" 
        aria-haspopup="true"
        aria-label="See ranking"
        onClick={handleClick}
      >
        See Ranking
      </Button>
      <div className="anchor" ref={menuAnchorRef} />
      <Menu
        id="polluters-menu"
        anchorEl={menuAnchorRef.current}
        keepMounted
        open={menuOpen}
        onClose={handleClose}
      >
        {getMenuItems()}
      </Menu>
    </div>
  );
};

export default PollutersMenu;
