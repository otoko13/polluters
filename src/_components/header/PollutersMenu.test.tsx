import { fireEvent, render } from '@testing-library/react';
import * as React from 'react';
import PollutersMenu, { IPollutersMenuProps } from './PollutersMenu';
import { IPolluter } from '../../model/Polluter';

let props: IPollutersMenuProps;

beforeEach(() => {
  props = {
      onHovered: jest.fn(),
      onNothingHovered: jest.fn(),
      onSelected: jest.fn(),
      polluters: [
        {
          rank: 1,
          name: 'Test1',
        } as IPolluter,
        {
          rank: 2,
          name: 'Test2',
        } as IPolluter,
      ],
  };
});

describe('snapshot', () => {
    it('should look sensible', () => {
        const { container } = render(<PollutersMenu {...props} />);
        expect(container.firstChild).toMatchSnapshot();
    });
});

describe('hovering', () => {
  it('should fire props.onHovered when a polluter is hovered', () => {
      const { getAllByRole, getByLabelText } = render(<PollutersMenu {...props} />);
      fireEvent.click(getByLabelText('See ranking'));
      fireEvent.mouseOver(getAllByRole('menuitem')[1]);
      expect(props.onHovered).toHaveBeenCalledWith(props.polluters[1]);
  });

  it('should fire props.onNothingHovered when a polluter is moused out of', () => {
      const { getAllByRole, getByLabelText } = render(<PollutersMenu {...props} />);
      fireEvent.click(getByLabelText('See ranking'));
      fireEvent.mouseOut(getAllByRole('menuitem')[1]);
      expect(props.onNothingHovered).toHaveBeenCalled();
  });
});

describe('selecting', () => {
  it('should fire props.onSelected when a polluter is clicked', () => {
      const { getAllByRole, getByLabelText } = render(<PollutersMenu {...props} />);
      fireEvent.click(getByLabelText('See ranking'));
      fireEvent.click(getAllByRole('menuitem')[1]);
      expect(props.onSelected).toHaveBeenCalledWith(props.polluters[1]);
  });
});