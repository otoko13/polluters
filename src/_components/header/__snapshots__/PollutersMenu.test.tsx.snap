// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot should look sensible 1`] = `
<div
  class="PollutersMenu"
>
  <button
    aria-controls="polluters-menu"
    aria-haspopup="true"
    aria-label="See ranking"
    class="MuiButtonBase-root MuiButton-root MuiButton-contained"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      See Ranking
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <div
    class="anchor"
  />
</div>
`;
